### PR TITLE
fix: dependent notifications on project while destroy

### LINKED ISSUE #3275: fix: ActionView::Template::Error: Couldn't find Project with 'id'
**Describe the bug**
New unhandled error in production from [circuitverse](https://app.bugsnag.com/circuitverse/circuitverse?i=sk&m=nw) in projects#destroy [(details)](https://app.bugsnag.com/circuitverse/circuitverse/errors/630f03a63d053400098bc0bf?event_id=630f03a6009657c426200000&i=sk&m=nw)
Unhandled error
ActionView::Template::Error: Couldn't find Project with 'id'=792042
Location
app/notifications/fork_notification.rb:8 - message

**Are you working on this issue? (Yes/No)**
Yes

### PR DESCRIPTION
**Pull request by vedant-jain03**
Fixes #3275 

Note: Please check **Allow edits from maintainers.** if you would like us to assist in the PR.

### CODE REVIEW

File: app/controllers/projects_controller.rb [hunk 1]
```diff
@@ -144,6 +145,15 @@ def sanitize_name
       params[:project][:name] = sanitize(project_params[:name])
     end
 
+    # destroy all dependent notifications
+    def filter_dependent_notifications
+      NoticedNotification.find_each do |notification|
+        if notification.params[:project_id] == @project.id
```
[the remark this answers is not in the document]
Comment (tachyons) [reply app/controllers/projects_controller.rb:151]: This is super inefficient, just add dependent: destroy to notifications associations in the project model

**Comment by vedant-jain03** [comment]
@tachyons Kindly Review
`dependent: true` default in `has_noticed_notification` helper association!

**Comment by vedant-jain03** [comment]
## What I did:
- Initially, we were storing `project_id` which was causing an error in destroying the dependent notifications on deleting the project, so now we are storing the object with the param name `project`.
- The data migration for old notifications has not been run yet, so modified that file too!
- Added new data migration for existing `noticed_notifications` where the notifications for which Project does not exist will get deleted and for existing will update the params with the `Project` and `user` object.

@tachyons @abstrekt @pavanjoshi914

File: app/controllers/users/noticed_notifications_controller.rb [hunk 2]
```diff
@@ -11,7 +11,7 @@ def index
   def mark_as_read
     notification = NoticedNotification.find(params[:notification_id])
     notification.update(read_at: Time.zone.now)
-    project = Project.find(notification.params[:project_id])
+    project = notification.params[:project]
```
Comment (tachyons) [inline app/controllers/users/noticed_notifications_controller.rb:14]: What is the difference does this change make ?

Comment (tachyons) [inline app/controllers/users/noticed_notifications_controller.rb:14]: What is the difference does this change make ?

Comment (vedant-jain03) [reply app/controllers/users/noticed_notifications_controller.rb:14]: So previously we were finding the record using `project_id` in `Project` but now the object is already saved in `params[: project]` so there is no need to find it and complicate it once again!

**Comment by vedant-jain03** [comment]
@tachyons we can merge this!
The error in code climate is recursive, even if we merge the `if` in a single line, the line is becoming long and it is suggested to use the `block` statement only!

**Comment by tachyons** [comment]
@vedant-jain03 Then disable specific rubocop cop with rubocop:disable command

**Comment by vedant-jain03** [comment]
@tachyons Kindly Review!

**Pull request merged**